compiler/rustc_session/src/config.rs

```diff
@@ -370,9 +370,14 @@ impl LinkSelfContained {
     /// components was set individually. This would also require the `-Zunstable-options` flag, to
     /// be allowed.
     fn are_unstable_variants_set(&self) -> bool {
-        let any_component_set =
-            !self.enabled_components.is_empty() || !self.disabled_components.is_empty();
-        self.explicitly_set.is_none() && any_component_set
+        if self.explicitly_set.is_some() {
+            return false;
+        }
+
+        // Only the linker component is stable, anything else is thus unstable.
+        let mentioned_components = self.enabled_components.union(self.disabled_components);
+        let unstable_components = mentioned_components - LinkSelfContainedComponents::LINKER;
+        !unstable_components.is_empty()
```

Comment:
I'd expect `-C link-self-contained=+/-linker` to also be stabilized only on `x86_64-unknown-linux-gnu`.
It certainly doesn't work correctly on mingw, and probably also on bare metal targets using `rust-lld` by default.

Reply:
Done as a separate commit to make review easier.

@rustbot ready